File: 51.n-queens.py
# @leet start
class Solution:

    def createBoard(self, n):
        return ["." * n for _ in range(n)]
    
    def cloneBoard(self, board):
        result = []
        for b in board:
            result.append([i for i in b])
        return result
    
    def is_valid_pos(self, x, y, board):
        n = len(board)
        for i in range(n):
            if i == x:
                continue
            if board[y][i] == "Q":
                return False

        for i in range(n):
            if i == y:
                continue
            if board[i][x] == "Q":
                return False

        i = x + 1
        j = y + 1
        while i < n and j < n:
            if board[j][i] == "Q":
                return False

        i = x - 1
        j = y + 1
        while i < n and j < n:
            if board[j][i] == "Q":
                return False

        i = x - 1
        j = y - 1
        while i < n and j < n:
            if board[j][i] == "Q":
                return False

        i = x + 1
        j = y - 1
        while i < n and j < n:
            if board[j][i] == "Q":
                return False
        
        return True

    def solveNQueens(self, n: int) -> List[List[str]]:
        result = []
        curr_board = self.createBoard(n)

        def dfs(curr_n, board, y_start):
            if curr_n <= 0:
                result.append(self.cloneBoard(board))

            for y in range(y_start, n):
                for x in range(n):
                    if board[y][x] == '.':
                        s_val = board[y]
                        s_new = s_val[:x] + "Q" + s_val[x+1:]
                        board[y] = s_new
                        dfs(curr_n - 1, board, y)
                        board[y] = s_val

        dfs(n, curr_board, 0)

        return result
        
# @leet end

File: 51.n-queens.cpp
// @leet start
class Solution {
public:
    vector<string> createEmptyBoard(int n) {
        vector<string> result;
        string str = "";

        for (int i = 0; i < n; i++) {
            str += ".";
        }

        for (int i = 0; i < n; i++) {
            result.push_back(str);
        }

        return result;
    }

    bool isValid(const vector<string> &board, int n, int x, int y) {
        for (int i = 0; i < n; i++) {
            if (board[y][i] == 'Q') {
                return false;
            }

            if (board[i][x] == 'Q') {
                return false;
            }
        }

        int i = x;
        int j = y;

        while (i < n && j < n) {
            if (board[j][i] == 'Q') {
                return false;
            }
            i++;
            j++;
        }

        i = x;
        j = y;
        while (i >= 0 && j < n) {
            if (board[j][i] == 'Q') {
                return false;
            }
            i--;
            j++;
        }

        i = x;
        j = y;
        while (i < n && j >= 0) {
            if (board[j][i] == 'Q') {
                return false;
            }
            i++;
            j--;
        }

        i = x;
        j = y;
        while (i >= 0 && j >= 0) {
            if (board[j][i] == 'Q') {
                return false;
            }
            i--;
            j--;
        }

        return true;
    }

    void dfs(vector<string> &board, int size, int start_x, int start_y, int left, vector<vector<string>> &result) {
        for (int y = start_y; y < size; y++) {
            for (int x = start_x; x < size; x++) {
                if (isValid(board, size, x, y)) {
                    if (left - 1 == 0) {
                        board[y][x] = 'Q';
                        vector<string> temp = board;
                        result.push_back(temp);
                        board[y][x] = '.';
                        continue;
                    }
                    board[y][x] = 'Q';
                    dfs(board, size, x, y, left - 1, result);
                    board[y][x] = '.';
                }
            }
        }
    }

    vector<vector<string>> solveNQueens(int n) {
        vector<vector<string>> result;
        vector<string> curr = createEmptyBoard(n);
        dfs(curr, n, 0, 0, n, result);
        return result;
    }
};
// @leet end
